### PR TITLE
Moving away from `Integer | Nothing` to `Rows_To_Read` for limiting number of rows.

### PR DESCRIPTION
**Pull request by jdunkerley**
### Pull Request Description

- Added a new `Rows_To_Read` type with conversions from `Nothing` and integers.
- Updated `read` on `Table`, `Column`, `DB_Table` and `DB_Column`.
- Updated `Delimited_Format.Delimited` to use `Rows_To_Read` for `row_limit`.
- Updated `Excel_Format.Sheet` and `Excel_Format.Range` to use `Rows_To_Read` for `row_limit`.
- Updated `Excel_Workbook.read` to use `Rows_To_Read`.
- Updated `Connection.read` (in all connection types) to use `Rows_To_Read`.

image

image

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: distribution/lib/Standard/Table/0.0.0-dev/src/Rows_To_Read.enso [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    ## PRIVATE
+    attach_warning self input:Table -> Table = case self of
+        Rows_To_Read.First_With_Warning rows -> if input.row_count <= rows then input else
+            Problem_Behavior.Report_Warning.attach_problem_after (input.take (First rows)) <|
+                Not_All_Rows_Downloaded.Warning rows
+        _ -> input
+
+## PRIVATE
+Rows_To_Read.from (that:Nothing) =
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Rows_To_Read.enso:47]: Are these here only to handle existing code that uses the old interface?

Comment (jdunkerley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Rows_To_Read.enso:47]: Yep exactly. Any old limits will then be ok. Though not if the warning flag was set.

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Table/0.0.0-dev/src/Rows_To_Read.enso [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+type Rows_To_Read
+    ## Read all rows.
+    All_Rows
+
+    ## Read the first specified number of rows.
+    First rows:Integer=1000
+
+    ## First 1000 Rows with an attached warning if more to read.
+       This is the default for database reading.
+    First_With_Warning rows:Integer=1000
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Rows_To_Read.enso]: 1000 is the default but can be changed, so maybe:
```suggestion
    ## Read a limited number of rows, attaching warning if there is more.
       By default, a 1000 rows are read.
       This is the default for database reading.
    First_With_Warning rows:Integer=1000
```

File: distribution/lib/Standard/Table/0.0.0-dev/src/Column.enso [hunk 3]
```diff
@@ -2140,13 +2141,11 @@ type Column
...
        Arguments:
        - max_rows: specifies the maximum number of rows to read.
-         If `Nothing`, all available rows are returned.
-       - warn_if_more_rows: if set to `True`, a warning is attached to the
-         result if the number of rows returned by the query exceeds `max_rows`.
-    read : (Nothing | Integer) -> Boolean -> Column
-    read self (max_rows : Nothing | Integer = Nothing) (warn_if_more_rows:Boolean = True) =
+    @max_rows Rows_To_Read.default_widget
+    read : Rows_To_Read -> Column
+    read self (max_rows : Rows_To_Read = ..All_Rows) =
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Column.enso:2146]: This `..All_Rows` looks so much clearer than `Nothing`. I'm really happy with our switching towards more meaningful arguments. And with autoscoping it is still quite terse :)

File: distribution/lib/Standard/Table/0.0.0-dev/src/Rows_To_Read.enso [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        Rows_To_Read.All_Rows -> Nothing
+        Rows_To_Read.First rows -> rows
+        Rows_To_Read.First_With_Warning rows -> rows
+
+    ## PRIVATE
+    attach_warning self input:Table -> Table = case self of
+        Rows_To_Read.First_With_Warning rows -> if input.row_count <= rows then input else
+            Problem_Behavior.Report_Warning.attach_problem_after (input.take (First rows)) <|
+                Not_All_Rows_Downloaded.Warning rows
+        _ -> input
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Rows_To_Read.enso:44]: Nice that this is now in a single place, makes the usages much clearer

**Review by radeusgd — APPROVED** [review]
Awesome

**Pull request merged**